lib/rig_inbound_gateway_web/vconnection.ex

```diff
@@ -0,0 +1,204 @@
+defmodule RigInboundGatewayWeb.VConnection do
+  @moduledoc """
+  The VConnection acts as an abstraction for the actual connections.
+
+  It monitors the processes, sends out heartbeats and subscription refreshes and sends out events to the serverless engine when intialized, connection is lost or connection is re-established.
+  """
+
+  alias Rig.EventFilter
+  alias Rig.RingBuffer
+  alias RigCloudEvents.CloudEvent
+  alias RigInboundGateway.Events
+
+  require Logger
+
+  use GenServer
+
+  # TODO: Discuss a reasonable kill delay
+  @vconnection_kill_delay 120_000
+
+  # TODO: Discuss a reasonable buffer size
+  @buffer_size 100
+
+  def start(pid, subscriptions, heartbeat_interval_ms, subscription_refresh_interval_ms) do
+    GenServer.start(__MODULE__,
+    %{
+      target_pid: pid,
+      subscriptions: subscriptions,
+      heartbeat_interval_ms: heartbeat_interval_ms,
+      subscription_refresh_interval_ms: subscription_refresh_interval_ms,
+      kill_timer: nil,
+      monitor: nil,
+      event_buffer: RingBuffer.new(@buffer_size)
+    })
+  end
+
+  def start(heartbeat_interval_ms, subscription_refresh_interval_ms) do
+    start(nil, [], heartbeat_interval_ms, subscription_refresh_interval_ms)
+  end
+
+  @impl true
+  def init(state) do
+    # TODO: Publish connect event
+
+    Logger.debug(fn -> "New VConnection initialized #{inspect(self())}" end)
+
+    # We register subscriptions:
+    send self(), {:set_subscriptions, state.subscriptions}
+
+    # We schedule the initial heartbeat:
+    Process.send_after(self(), :heartbeat, state.heartbeat_interval_ms)
+
+    # And the initial subscription refresh:
+    Process.send_after(self(), :refresh_subscriptions, state.subscription_refresh_interval_ms)
+
+    if state.target_pid do
+      {:ok, create_monitor(state)}
+    else
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_call(:reconnect, from, state) do
+    if state.kill_timer do
+      Process.cancel_timer(state.kill_timer)
+      Logger.debug(fn -> "Kill-timer #{inspect(state.kill_timer)} stopped" end)
+    end
+
+    if state.monitor do
+      # Demonitor old connection in case we replace an existing connection
+      Process.demonitor(state.monitor)
+    end
+
+    # In case an old connection exists, kill it off before replacing it
+    send! state.target_pid, :close
+
+    {pid, _} = from
```

Comment:
calling everything `pid` makes the code rather hard to read. what about `{from_pid, _} = from`?